crates/primitives/src/forkid.rs

```diff
@@ -207,8 +259,16 @@ impl ForkFilter {
                 return Ok(())
             }
 
+            // We check if this fork is time-based or block number-based
+            // NOTE: This is a bit hacky but I'm unsure how else we can figure out when to use
+            // timestamp vs when to use block number..
+            let head_block_or_time = match self.cache.epoch_start {
+                ForkFilterKey::Block(_) => self.head.number,
+                ForkFilterKey::Time(_) => self.head.timestamp,
+            };
+
             //... compare local head to FORK_NEXT.
-            return if self.head >= fork_id.next {
+            return if head_block_or_time >= fork_id.next {
```

Comment:
The type information is lost from the EIP eh? so we really need to make an assumption that when `head_block_or_time` is `ForkFilterKey::Block` that `fork_id.next` is gonna correspond to a block value, and correspondingly for `Time`?